### ISSUE TITLE
When DynamoDB DescribeEndpoints is used, wrong scheme may be tacked on the result

### ISSUE THREAD

**Issue by nyh** [aws/aws-sdk-cpp#2554]
### Describe the bug

I saw this problem in the context of the new AWS DynamoDB Shell (https://github.com/awslabs/dynamodb-shell), which is written using the AWS C++ SDK.

I ask it to use a specific endpoint URL, e.g., `http://localhost:8000` (i.e., a local installation such as DynamoDB Local or ScyllaDB Alternator).

It turns out that the C++ SDK doesn't use this endpoint URL immediately, but rather tries to run a `DescribeEndpoints` request on this URL, to get a better (?) endpoint. The relevant code looks like this (it's repeated many times in generated code):

```cpp
      DescribeEndpointsRequest endpointRequest;
      auto endpointOutcome = DescribeEndpoints(endpointRequest);
      if (endpointOutcome.IsSuccess() && !endpointOutcome.GetResult().GetEndpoi
nts().empty())
      {
        const auto& item = endpointOutcome.GetResult().GetEndpoints()[0];
        ...
        endpoint = Aws::String(SchemeMapper::ToString(m_clientConfiguration.scheme)) + "://" + item.GetAddress();
```

The problem here is that no matter what the original endpoint URL was, the new endpoint built here will always use some default scheme (and that defaults to "https"), which in this example is incorrect - the local installation used http, not https.

Note that this problem is not reproducible in DynamoDB Local, because it appears that (I don't know why) that implementation simply doesn't implement the `DescribeEndpoints` request - so the SDK skips the attempt to rediscover the endpoint. But ScyllaDB Alternator does implement `DescribeEndpoints`, so this bug is exposed.

### Expected Behavior

The C++ SDK should either trust the given endpoint (like `http://localhost:8000`) without second-guessing it with a `DescribeEndpoints` request, or if `DescribeEndpoints` is used, it should take the scheme not from some default configuration (namely, "https") but rather from the original endpoint (in this example, "http").

### Current Behavior

When running `DDBSH_ENDPOINT_OVERRIDE=http://localhost:8000 ddbsh` against ScyllaDB Alternator serving DynamoDB API on http://localhost:8000, we get the following output:

```
ddbsh - version 0.6.1
[ddbsh.cpp, 188], Connection Failed. . . 
	curlCode: 35, SSL connect error
us-east-1 (*)> 
```

Note the bizarre "SSL connect error" coming from an http, not https endpoint. With wireshark, we can see what really happened. The C++ SDK sent a `DescribeEndpoints` to http://localhost:8000 and it responded with `localhost:8000`. The C++ SDK then tacked on the beginning an "https://" (the default scheme in the default client configuration), and came up with a new endpoint URL `https://localhost:8000`, so it then started sending SSL requests on port 8000, which is an unencrypted HTTP server, so it didn't work (wireshark shows that the broken request is repeated multiple times, until finally giving many seconds later and producing this SSL connect error.

### Reproduction Steps

See details above.
As I noted above, this issue cannot be reproduced on DynamoDB Local because they chose not to implement the DescribeEndpoints requests, so the broken scheme logic is skipped. But it can reproduced on ScyllaDB Alternator which does implement DescribeEndpoints. I guess it can also be reproduced on AWS DynamoDB if a user deliberately chooses to use the http scheme but this code will revert it to use https instead (the result will be working, but use an unintended scheme).

### AWS CPP SDK version used

1.11.107

### Compiler and Version used

gcc (GCC) 12.3.1

### Operating System and version

Linux, Fedora 37

**Comment by nyh** [comment]
By the way, looking more closely at the SDK code (https://github.com/aws/aws-sdk-cpp/blob/main/generated/src/aws-cpp-sdk-dynamodb/source/DynamoDBClient.cpp), it seems that if the `DescribeEndpoints` request fails (as I believe is the case in DynamoDB Local), it is retried before every single request. This too, I think, is a bug. It makes sense to cache (even if for a relatively short duration) the knowledge that the request failed instead of retrying it before each and every single request.

**Comment by yasminetalby** [comment]
Hello @nyh ,

Thank you very much for your detailed submission. We really appreciate you sharing your feedback and work.
You are right the SDK should be using the endpoint provided in the endpoint override. It would also be a better behavior to not have to change the default scheme through the client configuration when an endpoint is provided through endpoint override.

I will work on a fix for this issue and will post an update here once it is released.

Thank you once again for your time and collaboration.

Sincerely,

Yasmine

**Comment by yasminetalby** [comment]
This should be fixed in [1.11.148](https://github.com/aws/aws-sdk-cpp/tree/1.11.148).

Thank you very much for bringing this up to our attention.

Best regards,

Yasmine
